### DISCUSSION TITLE
how this should work?  image captcha outside e107 root

### DISCUSSION THREAD

**Discussion by Jimmi08** [e107inc/e107#5075] [Q&A Core]
Hi,
I did this in e107 root.
File test.php
```
<?php
require_once('class2.php'); 
echo e107::getSecureImg()->renderimage();

```
It works, image is rendered. 

But if I put e107 in a subdirectory and I do the same thing in root of site, 

```
<?php
require_once('e107/class2.php');
echo e107::getSecureImg()->renderimage();
```
there is Render Failed message.

What I am missing?  Is this possible? If not, why? 

I found one reason:
https://github.com/e107inc/e107/issues/5074

Thanks

**Comment by Deltik** [comment]
Two sessions are being used: One to generate the image, and a different, unrelated one to retrieve a CAPTCHA image that doesn't exist in that session.

The reason is that the first cookie is being set at the parent level and the second cookie is scoped for the e107 subdirectory:

![image](https://github.com/e107inc/e107/assets/1364268/9e1ccd35-13be-4a4f-b3e1-a7e643dd80d1)

The CAPTCHA is one of my biggest session pet peeves because it pollutes the session storage with thousands or millions of one-time guests, depending on the number of visitors.  I want to fix your use case by replacing the sessions with [RFC 7519](https://datatracker.ietf.org/doc/html/rfc7519) tokens (known as JSON Web Tokens). The main hurdle is the encryption key. I don't know where to get a safe encryption key that only the e107 server knows, since e107 doesn't come with one.

**Reply by Deltik** [reply]
@Jimmi08: I've rewritten the CAPTCHA answer storage system in https://github.com/e107inc/e107/pull/5541 so that PHP sessions are no longer used.  Could you try [this patch](https://patch-diff.githubusercontent.com/raw/e107inc/e107/pull/5541.patch) to see if your problem persists?

**Comment by Alex-e107nl** [comment]
I saw this: [https://codepen.io/hirensojitra/pen/OJBpVvB](https://codepen.io/hirensojitra/pen/OJBpVvB) a simple clearly readable solution for the default captcha, could that be a solution?

**Reply by Deltik** [reply]
Such readable CAPTCHAs are trivially readable by bots these days.  Bots are now better at solving human-oriented puzzles than humans, so we can't really do much than deter the most basic automation. 😕

**Reply by Jimmi08** [reply]
What do you think about this solution?  I just saw it in theme I am working on:

Added
`<input id="form_botcheck" name="form_botcheck" class="form-control" type="hidden" value="">`

and then:
```
 $botcheck = $_POST['form_botcheck'];
 if( $botcheck == '' ) {
 $message = 'Bot <strong>Detected</strong>.! Clean yourself Botster.!';
 $status = "false";
and email is not sent
}

```

Just opinion, I don't ask to change something.  I remember that this was a technique 20 years ago. It surprised me when I noticed it  now.

**Reply by Alex-e107nl** [reply]
I understand that it is complicated, but if you do not want to be dependent on an external party, there should be a reasonably good solution, and a captcha that shows characters that ordinary people cannot read is not a solution either. Jimmi08 that's a spam trap can we just build that into a theme? It can probably be avoided, but we can always try!

**Reply by Moc** [reply]
@Jimmi08 That technique is already used in signup in e107 core. It has an extra hidden field that should not be filled in (but it is filled in by bots).

**Reply by Alex-e107nl** [reply]
So.... why not in the contactform, is it difficult?

**Reply by Jimmi08** [reply]
@Moc what one? The only hidden field in form on signup.php is e_token. Bootstrap5 theme is used.

**Reply by Jimmi08** [reply]
I just remember how we did it those years ago.
- you add field "email" for bots, hidden - the best with CSS only, not required
- you add field with other name for user to enter email (like name  = "realfield" or "color")

After sending you checked  "email" field , you blocked the filled one and before saving to database you set email = realfield. 

Something like that. type email wasn't used either.

**Reply by Moc** [reply]
@Jimmi08 Correct, it was "email2" in /signup.php I think. But it appears to have been removed for some reason. Not sure why There's still a check in signup.php but I don't see where the hidden field is inserted. 

```php
if(!empty($_POST['email2'])) // spam-trap.
{
	exit; 	
}
```

**Reply by Jimmi08** [reply]
@Moc  it was in captcha in 1.0.4
```
	/**
	 * Render Img Tag
	 */
	function r_image()
	{
		$code = $this->create_code();
		return "<img src='".e_HTTP.$this->HANDLERS_DIRECTORY."secure_img_render.php?{$code}' alt='' />
		<input type='text' id='email2' name='email2' value='' style='display:none' />";
	}
```
